feat: version compatible with vue 3

elevatebart opened the pull request:
Hello @koca ,

Here is the first draft of the prism editor working with vue 3.
I have been asked to startup a vue-live engine for vue 3.

Would you be open to delivering this one on the `@next` channel?

Do you need anything more?

koca commented:
Hey @elevatebart! Thank you for the PR ♥️ i will review this after work(today).

elevatebart commented:
I just found out that the `@testing-library/vue` was indirectly dependent on vue 2 (through their version of the vue test utils).
It means that the unit tests will not work until they deliver their upgrade. 

Could we to deliver this version as an alpha?

koca commented:
Why do i see my old commits? There should be only 1 commit i think 🤔

elevatebart commented:
Yep @koca I forgot to mention that.

Note that my PR is based on the `next` branch.

The `next` branch is behind `master` by those 20 commits.
I started by rebasing `master` into `next` before starting the refactor.

The only change that matters here is the last commit:

https://github.com/koca/vue-prism-editor/pull/95/commits/394b1ce6f827e9b2d262fea84049363ba9188bee

koca commented:
can you rebase this branch and update the PR?

koca commented:
oh i see its not from master branch

koca commented on packages/vue-prism-editor/src/Editor.ts:
```diff
@@ -92,9 +93,9 @@ export const PrismEditor = Vue.extend({
+    modelValue: {
       immediate: true,
-      handler(newVal: string): void {
+      handler(this: any, newVal: string): void {
```
why did you add `this:any` to some methods?

koca commented on packages/vue-prism-editor/src/Editor.ts:
```diff
@@ -104,31 +105,31 @@ export const PrismEditor = Vue.extend({
     content: {
       immediate: true,
-      handler(): void {
+      handler(this: any): void {
```
same `this:any`

koca commented on packages/vue-prism-editor/src/Editor.ts:
```diff
@@ -1,17 +1,18 @@
-const KEYCODE_BRACKETS = 219;
-const KEYCODE_QUOTE = 222;
-const KEYCODE_BACK_QUOTE = 192;
-const KEYCODE_ESCAPE = 27;
+const KEYCODE_ENTER = 'Enter';
+const KEYCODE_TAB = 'Tab';
+const KEYCODE_BACKSPACE = 'Backspace';
+const KEYCODE_Y = 'y';
+const KEYCODE_Z = 'z';
+const KEYCODE_M = 'm';
+const KEYCODE_PARENS = '(';
+const KEYCODE_BRACKETS = '{';
+const KEYCODE_QUOTE = '"';
+const KEYCODE_BACK_QUOTE = '`';
+const KEYCODE_ESCAPE = 'Escape';
```
these are not keycodes anymore what was the problem?

elevatebart replied:
`event.keyCode` is deprecated
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode

and `event.key` seems compatible with IE9+ 
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key

It has nothing to do with vue 3 and I should remove this change from this PR.

koca replied:
i think its okay we can keep the changes

elevatebart replied:
<img width="378" alt="Screen Shot 2020-10-16 at 2 37 44 PM" src="https://user-images.githubusercontent.com/5592465/96301729-38bbcf00-0fbd-11eb-816a-780bd8acf47b.png">

elevatebart commented:
Thank you @koca for the comments, 

`this: any` turned out to be unnecessary so I removed it.

the keyCode property is deprecated but should not prevent vue.3 from working so I removed this change too.

I hope it simplifies the review a little.

Sorry again for the confusion.

elevatebart commented on packages/vue-prism-editor/src/Editor.ts:
```diff
@@ -55,7 +56,7 @@ export const PrismEditor = Vue.extend({
       default: false,
     },
-    value: {
+    modelValue: {
```
value => modelValue
https://v3.vuejs.org/guide/migration/v-model.html#migration-strategy

elevatebart commented on packages/vue-prism-editor/src/Editor.ts:
```diff
@@ -261,7 +262,7 @@ export const PrismEditor = Vue.extend({
       input.selectionEnd = record.selectionEnd;
 
-      this.$emit('input', record.value);
+      this.$emit('update:modelValue', record.value);
```
`this.$emit('input' => 'update:modelValue'`

https://v3.vuejs.org/guide/migration/v-model.html#migration-strategy

elevatebart commented on packages/vue-prism-editor/src/Editor.ts:
```diff
@@ -306,13 +307,11 @@ export const PrismEditor = Vue.extend({
-        // onKeyDown(e);
-        this.$emit('keydown', e);
+      // onKeyDown(e);
+      this.$emit('keydown', e);
```
`this.$listeners` has been removed in vue 3 (or is null as far as I can tell) which triggers an undefined error.

elevatebart replied:
After investigation, `this.$listeners` exists only if it has been implemented, with an `@click=` or an `@keydown=` 
It should be protected against null.
Since it is deprecated anyway, I removed it completely.

koca replied:
yep they've merged `$listeners` to `$attrs`. So `$attrs` has both `$attrs` and `$listeners` in vue 3

elevatebart commented on packages/vue-prism-editor/src/Editor.ts:
```diff
@@ -512,27 +511,23 @@ export const PrismEditor = Vue.extend({
     },
   },
-  render(h): VNode {
+  render() {
```
h is now imported from vue

elevatebart commented on packages/vue-prism-editor/src/Editor.ts:
```diff
@@ -512,27 +511,23 @@ export const PrismEditor = Vue.extend({
       'div',
       {
-        staticClass: 'prism-editor__line-numbers',
+        class: 'prism-editor__line-numbers',
```
h has changed

https://v3.vuejs.org/guide/migration/render-function-api.html#vnode-props-format

elevatebart commented:
I have added comments in this commit to explain my changes

https://github.com/koca/vue-prism-editor/pull/95/commits/77bb9af05a21e1536add2233192fef0e67c50e9c

koca commented:
There is only console.log left can you please remove it and i've tested on my local, seems to be working fine. We can't run tests but I'll publish the alpha version for you. Thank you @elevatebart 😊👍

koca commented:
Hey @elevatebart just released `v2.0.0-alpha.1`

example vue 3 csb: https://codesandbox.io/s/vue-3-support-for-vue-prism-editor-fm7v5

install:

```sh
npm install vue-prism-editor@alpha
```

or

```sh
yarn add vue-prism-editor@alpha
```

cheers 🎉
